com.github.marinm.songrec.json

```diff
@@ -0,0 +1,47 @@
+{
+    "app-id": "com.github.marinm.songrec",
+    "runtime": "org.freedesktop.Platform",
+    "runtime-version": "20.08",
+    "sdk": "org.freedesktop.Sdk",
+    "sdk-extensions" : [
+        "org.freedesktop.Sdk.Extension.rust-stable"
+    ],
+    "command": "songrec",
+    "finish-args": [
+        "--share=network",
+        "--share=ipc",
+        "--socket=fallback-x11",
+        "--socket=wayland",
+        "--device=dri",
+        "--filesystem=host:ro",
+        "--filesystem=/tmp:ro",
```

Comment:
Why read access to those?

Reply:
It's for the file picker of the app to work. I personally often use `/tmp` for storing things (such as video soundtracks downloaded using `youtube-dl`...).

Reply:
In general, we vastly prefer permissions to be locked down by default, so here I'd say `xdg-music` and `xdg-download` would be sufficient. It's easy to change via CLI or [GUI](https://flathub.org/apps/details/com.github.tchx84.Flatseal).

Reply:
That being said, any reason you are not using `GtkFileChooserNative` here? Looks like it operates only on single files (for both sound files and CSV export). This way you could drop all filesystem permissions and let the portal handle it.

Reply:
I didn't know about it. I could, but won't the command-line interface that directly accepts a sound file path from the user be broken?

Reply:
Yeah, but while I don't have any data for this, I don't think many Flathub users run apps from terminal.

Reply:
I think however that the command line input is an important feature of the app (for example, to enable batch recognition, to the extent accepted by the law) and that users wouldn't be pleased to change distribution channel to use the full version of the app.

Reply:
Something like `flatpak run --file-forwarding appid @@ %f @@` also works from CLI via portals. Ultimately, it's of course your choice as the developer, however `host` and `/tmp` are still too wide. When these are limited, at least two `home:ro`, LGTM.

Reply:
I'm not sure how I could limit `/tmp` to `home`? Should users explicitly be warned not to store things in `/tmp` (which is a common practice especially when using tools such as `youtube-dl` which don't have commonly used GUI frontends) when installing the app through the Flatpak distribution channel?

Reply:
I mean, it is what it is. We as Flathub decided it's better to ship software with limited but not unexpected permissions instead of just adding `--filesystem=host` by default. I'm not denying it's not inconvenient, but we need to draw the line somewhere. I still think that outside of blogs or project website, majority of users will come from https://flathub.org and GNOME/KDE frontends, and so they will never know it's usable from CLI.

Reply:
I guess then that the way to incur the less UX degradation patterns would be implement using `GtkFileChooserNative` for the graphical file picker, and keep a read access to `home` for the command line.

However, when using the `show()` method on `GtkFileChooserNative` (from [`gtk::NativeDialogExt::show`](http://gtk-rs.org/docs/gtk/trait.NativeDialogExt.html#tymethod.show)) the dialog shows a short instant and almost immediately hides, while this is not what the docs says ("_When the user accepts the state of the dialog_ the dialog will be automatically hidden and the “response” signal will be emitted.").

Should this be considered as abnormal behavior and first reported to `gtk-rs`?

Reply:
Yeah, that sounds wrong. cc @bilelmoussaoui as I don't know Rust.

Reply:
For the time being it's fine to go with `--filesystem=home` so the submission is not blocked.

Reply:
> I guess then that the way to incur the least UX degradation patterns would be to implement using `GtkFileChooserNative` for the graphical file picker, and keep a read access to `home` for the command line.
> 
> However, when using the `show()` method on `GtkFileChooserNative` (from [`gtk::NativeDialogExt::show`](http://gtk-rs.org/docs/gtk/trait.NativeDialogExt.html#tymethod.show)) the dialog shows a short instant and almost immediately hides, while this is not what the docs says ("_When the user accepts the state of the dialog_ the dialog will be automatically hidden and the “response” signal will be emitted.").
> 
> Should this be considered as abnormal behavior and first reported to `gtk-rs`?

There's probably something you have done wrong. It's very easy to break things with dialogs in general with gtk3. Do you have your code that uses where I can have a look?

Also, gtk application supports opening a specific file, from cli without any file system access. (that's what's run when you are asked to open a file with a specific app)

Reply:
The full content of my input dialog event handler after the modifications is here:

```rust
            let file_chooser = gtk::FileChooserNative::new(
                Some("Select a file to recognize"),
                Some(&window),
                gtk::FileChooserAction::Open,
                Some("_Open"),
                Some("_Cancel")
            );
            
            let file_filter = gtk::FileFilter::new();
            
            file_filter.add_pattern("*.mp3");
            file_filter.add_pattern("*.wav");
            file_filter.add_pattern("*.flac");
            file_filter.add_pattern("*.ogg");
            
            file_chooser.set_filter(&file_filter);
            
            let processing_tx_3 = processing_tx.clone();

            file_chooser.connect_response(move |file_chooser, response| {
                if response == ResponseType::Ok {
                    recognize_file_button.hide();
                    
                    spinner.show();
                    
                    let input_file_path = file_chooser.get_filename().expect("Couldn't get filename");
                    let input_file_string = input_file_path.to_str().unwrap().to_string();
                    
                    processing_tx_3.send(ProcessingMessage::ProcessAudioFile(input_file_string)).unwrap();
                }
            });
            file_chooser.show();
```

Put back in context it's from [here](https://github.com/marin-m/SongRec/blob/0.1.1/src/gui/main_window.rs#L189)

Reply:
I replaced it with the blocking method `run()` instead of the non-blocking method `show()` (which shouldn't be much a problem as the app is multi-threaded) and now this works better. I also restricted the access in the manifest as requested.

Reply:
@marin-m your code is a bit wrong, the file path isn't something you should use inside the sandbox as the files you have access to are mounted inside the container as a completely different path. You should use the file uri instead. By using dialog.get_file, you should have a gio::File that would allow you to read/write into the file if needed.

Reply:
I tried to update my code to use `gio::File::new_for_uri` an `gio::File::new_for_path` and `gio::FileExt::load_contents` where relevant, but it gets tricky when it comes to the fact that the audio format decoding library that I use (whose entry point [`rodio::decoder::Decoder::new`](https://docs.rs/rodio/0.12.0/rodio/decoder/struct.Decoder.html)) requires an object implementing both the `Read` trait and the `'static` lifetime, which a native Rust `std::fs::File` seems to implement but a dynamically obtained buffer obtained from GIO does not.

I'm not sure if I can adapt my code without changing the audio decoding library then.

Reply:
You can get a stream from your gio file instead of using a load_contents, the input stream implements Read I guess.

Reply:
Not according to the doc: http://gtk-rs.org/docs/gio/struct.FileInputStream.html

The `read()` method on it is implemented by `InputStreamExtManual` which is a distinct trait. http://gtk-rs.org/docs/gio/prelude/trait.InputStreamExtManual.html

Reply:
I think that I could hack this out through reading the file with GIO, then writing to a temporary file, and passing it back to the library, but I don't know whether this is needed as trying the Flatpak build by the buildbot with the current code (https://github.com/flathub/flathub/pull/1914#issuecomment-720511946) and using it to loa a file from `/tmp` or the home directory through the GUI works.

So adapting the code for using URIs is not needed I guess?

Reply:
@marin-m, fileInputStream does implement the ExtManual trait (it's just a manual trait that contains code that wasn't generated automatically from the gir annotations), you just need to include it with `use gio::InputStreamExtManual`

Reply:
Yes, but this shouldn't make Rust think that the trait implements `Read`, except if I make a `Read` wrapper myself. Anyway, reading the file-picker selected files inside Flatpak can be observed to work without URIs or GIO, so I guess that I can avoid using these with the sale results and simplify the code.

Reply:
Well I don't follow here anymore. If your library takes something that implement Read, InputStream should work.

Reply:
Again, it seems that it clearly doesn't implement `Read` (whichever `Ext` trait I include or not, as the `read()` function implemented by `InputStreamExtManual` doesn't have the same prototype as the one implemented by `std::io::Read` anyway):

```rust
error[E0277]: the trait bound `gio::FileInputStream: std::io::Read` is not satisfied
  --> src/fingerprinting/algorithm.rs:48:72
   |
48 |         let converted_file = rodio::source::UniformSourceIterator::new(decoder, 1, 16000);
   |                                                                        ^^^^^^^ the trait `std::io::Read` is not implemented for `gio::FileInputStream`
   |
   = note: required because of the requirements on the impl of `std::iter::Iterator` for `rodio::Decoder<gio::FileInputStream>`

error[E0277]: the trait bound `gio::FileInputStream: std::io::Seek` is not satisfied
  --> src/fingerprinting/algorithm.rs:48:72
   |
48 |         let converted_file = rodio::source::UniformSourceIterator::new(decoder, 1, 16000);
   |                                                                        ^^^^^^^ the trait `std::io::Seek` is not implemented for `gio::FileInputStream`
   |
   = note: required because of the requirements on the impl of `std::iter::Iterator` for `rodio::Decoder<gio::FileInputStream>`
```

But that using GIO isn't needed to have the application working at all, as the currently produced Flatpak snapshot fully works for the GUI part, including file picking from various parts of the filesystem. So I'm not sure why the current code would be "wrong" or would need to use GIO or URIs.

Reply:
https://gtk-rs.org/docs/gio/prelude/trait.InputStreamExtManual.html#method.into_read this gives you something that implements std::io::Read.

Reply:
because you're still assuming that you have permission to every file out there. while the home permission works fine for those files in home, it won't work for any file outside of that directory.

Reply:
>https://gtk-rs.org/docs/gio/prelude/trait.InputStreamExtManual.html#method.into_read this gives you something that implements std::io::Read.

Ok, but the audio file format library also requires the concerned object to the thread-safe (to implement the `Send` trait:)

```rust
error[E0277]: `std::ptr::NonNull<glib::object::GObject>` cannot be sent between threads safely
  --> src/fingerprinting/algorithm.rs:45:23
   |
45 |         let decoder = rodio::Decoder::new(gio_file.read::<gio::Cancellable>(None)?.into_read())?;
   |                       ^^^^^^^^^^^^^^^^^^^ `std::ptr::NonNull<glib::object::GObject>` cannot be sent between threads safely
   |
   = help: within `gio::InputStreamRead<gio::FileInputStream>`, the trait `std::marker::Send` is not implemented for `std::ptr::NonNull<glib::object::GObject>`
   = note: required because it appears within the type `glib::object::ObjectRef`
   = note: required because it appears within the type `gio::FileInputStream`
   = note: required because it appears within the type `gio::InputStreamRead<gio::FileInputStream>`
   = note: required by `rodio::Decoder::<R>::new`
```

>because you're still assuming that you have permission to every file out there. while the home permission works fine for those files in home, it won't work for any file outside of that directory.

It seems that it will, at least on `/tmp` which is located outside of the home directory, as it can be observed by going through the following testing procedure:

1. Uninstall any former installation: `flatpak uninstall com.github.marinm.songrec -y`
2. Locally install the latest build: `flatpak install --user https://dl.flathub.org/build-repo/30447/com.github.marinm.songrec.flatpakref -y`
3. Put any decodable sound file (.WAV, .FLAC, .MP3...) in your `/tmp` directory
4. Run the application: `flatpak run com.github.marinm.songrec`
5. Click "Recognize from file"
6. Use the file picker to select the sound file that you have put in `/tmp` (outside of the home directory)

The current Flatpak snapshot (which is limited to `home:ro` by its permissions for reading, as it can be verified when trying to recognize a file from the command line) will then correctly work.